### PR TITLE
Migrate backward

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #985
* __->__ #980
* #976
* #975
* #972
* #971
* #974
* #967
* #966
* #964
* #963
* #962
* #961
* #960
* #959

### CODE REVIEW

File: test/test_bwd.py [hunk 1]
```diff
@@ -77,22 +65,28 @@ def run_worker(args):
...
     if args.rank == 0:
-        out = stage(x)
+        schedule.step(x)
     elif args.rank == args.world_size - 1:
-        out = stage(y)
+        losses = []
+        out = schedule.step(targets=targets, losses=losses)
     else:
-        stage()
+        schedule.step()
```
Comment (kwen2501) [inline test/test_bwd.py:78]: @H-Huang @wconstab Please check this new UI for training.

**Review by H-Huang — APPROVED** [review]

File: pippy/PipelineSchedule.py [hunk 2]
```diff
@@ -507,17 +569,26 @@ def backward_one_chunk(self, **kwargs) -> None:
...
-        return self.outputs[0].mean()
-
 
 class PipelineSchedule(ABC):
-    def __init__(self, stage: PipelineStageBase, n_microbatches: int):
+    def __init__(
+        self,
+        stage: PipelineStageBase,
+        n_microbatches: int,
+        loss_fn: Optional[Callable] = None,
```
[the remark this answers is not in the document]
Comment (kwen2501) [reply pippy/PipelineSchedule.py:574]: I agree loss_fn should not be an argument for step function. So I am giving it to the constructor of schedule. The main idea being: 
(1) let's leave the stages alone, because loss_fn is not part of the model in this case; thus we should try to make stages loyal to the original model. 
(2) calculating loss is usually in the training step, hence it is put in schedule now. (schedule is a pre-defined pattern for training step).
(3) Of course, if model is defined with loss inside, we should respect that too, as in tracer path's previous implementation. I am dropping this support in this first implementation, but will add it back after revisiting model definitions on HF.

File: test/test_bwd.py (hunk 1 above)
Comment (kwen2501) [reply test/test_bwd.py:78]: The return format of `(out, losses)` would seem a bit "arbitrary" to me, meaning, user would need to read closely our doc to know what the format would be. And it could make the API harder to change in future (output is harder to extend/change than argument list).

File: pippy/PipelineSchedule.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:574]: I was saying loss_fn could potentially be an argument in the `step()` function.
1) Agree that loss isnt related to the stages / model
2) schedule doesn't necessarily need to be only for training, you could (and we can provide a default) for a schedule with only forward passes for inference? Not sure if that is overkill

File: pippy/PipelineSchedule.py [hunk 3]
```diff
@@ -599,15 +702,33 @@ def step_microbatches(
...
+            assert isinstance(
+                losses, list
+            ), f"losses must be a list but got a {type(losses)}"
+            # Clean external container first
+            losses.clear()
+            # Copy internal losses to external container
+            losses.extend(internal_losses)
+
+    def step(
+        self, *args, targets=None, losses: Optional[List] = None, **kwargs
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py]: On second thought, I am okay with this. I think the only con I thought of passing losses is the user needs to know that this variable is going to be overridden at each step. I was also worried that losses need to be passed in even if the user does not want the losses but as long as the losses is optional this is okay.

File: test/test_bwd.py (hunk 1 above)
Comment (H-Huang) [reply test/test_bwd.py]: yeah since we are keeping the losses as arguments in the step function, we can just return the a single `out`, makes sense to me!

File: pippy/PipelineSchedule.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwen2501) [reply pippy/PipelineSchedule.py:574]: Re 2:
Right. In the inference case, user will not/need not to pass the `loss_fn` to schedule. `loss_fn` is Optional.
In fact, in the code, I check whether `loss_fn is None` to decide whether this schedule should do backward or not.

[the remark this answers is not in the document]
Comment (kwen2501) [reply pippy/PipelineSchedule.py:574]: I am not sure if user would have need to change `loss_fn` between steps. If they really have such need, they can recreate a schedule.

File: pippy/PipelineSchedule.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kwen2501) [reply pippy/PipelineSchedule.py]: Yeah, we should document that the `losses` container will be rewritten.
